### PR TITLE
first implementation of Conv1D from EJ

### PR DESCRIPTION
**Pull request by jmduarte**
Related to issue #13  

Took @ejk43's implementation of Conv1D layer from https://github.com/Xilinx/RFNoC-HLS-NeuralNet/blob/master/rfnoc/hls/nnet_lib/nnet_conv.h

Things changed:
- removed `hls::streams`
- changed output logic so output size is the same as input size (zero padding added only to the left of inputs)

Things to update/make configurable:
- configurable "zero" padding: valid, same, full, etc.
- number of filters
- reuse factor

@benjaminkreis also has some ideas about trying to make everything more parallelizable

### CODE REVIEW

**Comment by nhanvtran** [comment]
is there some code for testing?

**Comment by jmduarte** [comment]
added the test project code at bf723b9

**Comment by ejk43** [comment]
> changed output logic so output size is the same as input size (zero padding added only to the left of inputs)

Does this represent the "same" padding? 

As a related question, what other requirements would you like to see from conv1d and conv2d, especially regarding items like number of input/output channels for the convolution? One thing I found earlier,  at least with the "full serial" implementation, was that more options means more ways the HLS directives can go wrong...

**Comment by pierinim** [comment]
In terms of inputs to the 1D, we could imagine passing a list of 10 particles, each consisting of four numbers (energy + the angles for direction and one ID number for the kind of particle). The same input file could be used for LSTM. I would start with a binary classifier, so that the number of output layers is minimal

**Comment by benjaminkreis** [comment]
@ejk43 this is "same" padding in the sent the the number of inputs equals the number of outputs, but padding the data with zeros only on the "left" side might not be the standard.  We'll have to look at TensorFlow etc. to see how they split the zeros between the left and right sides.

I leave it to @jmduarte to answer about channels etc.

**Comment by benjaminkreis** [comment]
Here's the answer about same padding in TF: "In this case, the bottom and right sides always get the one additional padded pixel." (https://www.tensorflow.org/versions/r0.12/api_docs/python/nn/convolution)

**Comment by benjaminkreis** [comment]
The HLS csynth runs fine, so I'm going to go ahead a merge this to serve as our starting point.  
@jmduarte figured out that for each filter index, convolutional layers sum over the channels and apply one bias.

**Comment by benjaminkreis** [comment]
Sorry, I clicked comment too soon.

The HLS csynth runs fine, so I'm going to go ahead a merge this to serve as our starting point.  

At least for Keras/Tensorflow, @jmduarte figured out that for each filter index, convolutional layers sum over the channels and apply one bias.  I'm going to work on making this change for a next PR, which will not only affect the internal math but also the output shape.  @ejk43, let us know if you have a different understanding from your own experience with Tensorflow!

**Comment by ejk43** [comment]
> convolutional layers sum over the channels and apply one bias (see here for e.g.).

Interesting-- It's been a while, but I think I was under the impression each channel operated independently-- I must have gotten confused somewhere. Good catch! This makes more intuitive sense; that each *filter* is independent, but the channels are combined (for example, multiple channels in a color image).  Is that a correct interpretation?

**Comment by benjaminkreis** [comment]
Yeah, I think that is right.  One thing I'm not sure about is how you interpret the output when you connect it to another convolutional layer.  I suppose the different filter outputs from the first layer become channels for the second layer?
